Topic/schedule pipelined

vspetrov opened the pull request:
## What
Adds pipelined schedule interface and the  corresponding implementation for TL/UCP SRA allreduce algorithm

## Why ?
Pipelining improves performance of large message collectives in multiple cases

## How ?
- Pipelined schedule data structure is introduced that runs a collective as a sequence of fragments
- The interface is generic: user (tl/ucp e.g.) passes a cb that allocates a schedule -> it defines an execution of a collective fragment
- Number of simultaneously allocated fragments - pipeline depth - can be varied. 
- Frag schedules are restarted automatically keeping the dependencies
- 2 types of pipelines are supported: parallel and sequential. In parallel pipeline the tasks at the same position in frag schedule can be started out of order, in sequential the order is fixed (needed for example when TL that is used in frag schedule can't support out of order tasks)

vspetrov commented:
bot:retest

Sergei-Lebedev commented on src/schedule/ucc_schedule_pipelined.c:
```diff
@@ -0,0 +1,216 @@
+    schedule->super.super.finalize = ucc_schedule_pipelined_finalize;
+    schedule->super.super.post     = ucc_schedule_pipelined_post;
+    frags                          = schedule->frags;
+    for (i = 0; i < n_frags; i++) {
```
check n_frags < max_frags, return error othrewise

vspetrov commented on src/schedule/ucc_schedule_pipelined.h:
```diff
@@ -0,0 +1,42 @@
+    ucc_schedule_frag_setup_fn_t frag_setup;
+} ucc_schedule_pipelined_t;
+
+ucc_status_t ucc_schedule_pipelined_init(
```
done

vspetrov commented:
@bureddy i've modified "sequential" pipeline to depend on "EVENT_STARTED" instead of completed (according to discussion in WG). @Sergei-Lebedev your comments are also addressed. so, plz review when you have time (and then i can post new hierarchical CL PR).

Sergei-Lebedev commented on test/gtest/core/test_allreduce.cc:
```diff
@@ -242,3 +242,48 @@ TYPED_TEST(test_allreduce, multiple_cuda_inplace) {
+    UccCollCtxVec ctxs;
+    std::vector<ucc_memory_type_t> mt = {UCC_MEMORY_TYPE_HOST};
+
+#ifdef HAVE_CUDA
```
ifdef isn't needed here, we don't use any cuda functions

vspetrov commented on src/components/tl/ucp/allreduce/allreduce_sra_knomial.c:
```diff
@@ -87,32 +135,74 @@ ucc_tl_ucp_allreduce_sra_knomial_init(ucc_base_coll_args_t *coll_args,
+    *pipeline_depth = ucc_min(*n_frags, cfg->allreduce_sra_kn_pipeline_depth);
+}
+
+ucc_status_t ucc_tl_ucp_allreduce_sra_knomial_start(ucc_coll_task_t *task)
```
i renamed finalize to ucc_tl_ucp_allreduce_sra_knomial_finalize. "knomial" better represents what the algorithm is. the fact that i can be pipelined does not matter imho. sra can be implemented with ring e.g.

bureddy commented on src/schedule/ucc_schedule_pipelined.h:
```diff
@@ -0,0 +1,66 @@
+
+typedef struct ucc_schedule_pipelined ucc_schedule_pipelined_t;
+
+#define UCC_SCHEDULE_PIPELINED_MAX_FRAGS 4
```
question: at UCC level.  it should it make sense to use term "FRAG",  is this works only for fragmnetation (or) can we be used for any generic task pipeline?

vspetrov replied:
I'm not sure what do you mean? What is pipelining if you don't fragment the collective?

vspetrov commented on src/components/tl/ucp/allreduce/allreduce_sra_knomial.c:
```diff
@@ -87,32 +122,93 @@ ucc_tl_ucp_allreduce_sra_knomial_init(ucc_base_coll_args_t *coll_args,
+                     ucc_dt_size(coll_args->args.src.info.datatype);
+    *n_frags = 1;
+    if (msgsize > cfg->allreduce_sra_kn_frag_thresh) {
+        int min_num_frags = msgsize / cfg->allreduce_sra_kn_frag_size;
```
you are right

vspetrov commented on src/components/tl/ucp/tl_ucp.c:
```diff
@@ -52,6 +52,35 @@ static ucc_config_field_t ucc_tl_ucp_lib_config_table[] = {
+    {"ALLREDUCE_SRA_KN_SEQUENTIAL", "0",
+     "Type of pipelined schedule for SRA knomial alg (sequential/parallel)",
+     ucc_offsetof(ucc_tl_ucp_lib_config_t, allreduce_sra_kn_seq),
+     UCC_CONFIG_TYPE_UINT},
```
agree. changed to UCC_CONFIG_TYPE_BOOL which taskes "yes/no".
